File: app/src/main/java/com/mycompany/myapp/ui/BaseActivity.java
package com.mycompany.myapp.ui;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.mycompany.myapp.app.ApplicationComponent;
import com.mycompany.myapp.app.HasComponent;
import com.mycompany.myapp.app.MainApplication;
import com.mycompany.myapp.monitoring.CrashReporter;

import javax.inject.Inject;

import hugo.weaving.DebugLog;
import timber.log.Timber.Tree;

public abstract class BaseActivity<T> extends AppCompatActivity implements HasComponent<T> {
    @Inject
    protected Tree logger;

    @Inject
    protected CrashReporter crashReporter;

    private T component;

    @DebugLog
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        component = buildComponent();
    }

    protected abstract T buildComponent();

    @Override
    public T getComponent() {
        return component;
    }

    protected ApplicationComponent getApplicationComponent() {
        return ((MainApplication) getApplication()).getComponent();
    }
}

File: app/src/main/java/com/mycompany/myapp/ui/BaseActivity.kt
package com.mycompany.myapp.ui

import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.Navigation
import com.mycompany.myapp.R
import com.mycompany.myapp.app.ApplicationComponent
import com.mycompany.myapp.app.MainApplication
import javax.inject.Inject
import kotlin.reflect.KClass

abstract class BaseActivity : AppCompatActivity() {
    @Inject lateinit var viewModelFactory: ViewModelProvider.Factory

    protected val appComponent: ApplicationComponent
        get() = (application as MainApplication).component

    fun <VM: ViewModel> getViewModel(viewModelClass: KClass<VM>): VM {
        return ViewModelProviders.of(this, viewModelFactory).get(viewModelClass.java)
    }

    override fun onSupportNavigateUp() =
            Navigation.findNavController(findViewById(R.id.nav_host_fragment)).navigateUp()
}
